scintillometry/fourier/base.py

```diff
@@ -0,0 +1,290 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import numpy as np
+import operator
+from functools import wraps
+
+
+def _check_fft_exists(func):
+    # Decorator for checking that _fft and _ifft exist.
+    @wraps(func)
+    def check_fft(self, *args, **kwargs):
+        if '_fft' in self.__dict__ and '_ifft' in self.__dict__:
+            return func(self, *args, **kwargs)
+        raise AttributeError('Fourier transform functions have not been '
+                             'linked; run self.setup first.')
+    return check_fft
+
+
+class FFTBase(object):
+    """Base class for all fast Fourier transforms.
+
+    Provides a `setup` method to store relevant transform information in the
+    `data_format`, `norm` and `axes` properties, which respectively contain the
+    shapes and dtypes of the FFT arrays, the axes over which to perform the
+    FFT, and the normalization convention.  Also provides an `fftfreqs` method
+    to return the sample frequencies of the FFT along one axis.
+
+    Actual FFT classes must define a forward and backward transform as `fft`
+    and `ifft`, respectively (see placeholders in this class for the format).
+
+    Currently does not support Hermitian FFTs (where frequency-domain data is
+    real-valued).
+    """
+
+    _data_format = {'time_shape': None,
+                    'time_dtype': None,
+                    'freq_shape': None,
+                    'freq_dtype': None}
+    _axes = None
+    _norm = None
+
+    def setup(self, time_data, freq_dtype, axes=None, norm=None,
```

Comment:
Even after looking through the rest, I don't understand why this isn't just `__init__`...